Standalone and Add-on merge

GeoDerp opened the pull request:
Related to https://github.com/davidusb-geek/emhass/issues/269

GeoDerp commented:
This is still in development, Its here for to show progress. Alot has changed, and will still need to be.
Will edit this comment later with more info.

GeoDerp commented:
I am currently unsure how we should approach the parameter naming conventions moving forward. 
This PR is currently set up with the EMHASS code as the original parameter naming convention, and the options.json the add-on naming conversation (options.json being converted in `build_parms` with the associations list. 
However, we could also change all the parameters in the options.json to match the original.
Moreover, we could change the code to match the options.json parameter naming.
Either way I'm planning on making a script that will help users to convert from one to the other for an easier transition.

davidusb-geek commented:
It would be probably best to change the code to match the options.json, those names are way more explicit

GeoDerp commented:
*work in progress*

GeoDerp commented:
![Screenshot 2024-08-22 at 23-06-52 EMHASS Energy Management Optimization for Home Assistant](https://github.com/user-attachments/assets/17397b77-e656-42f3-8f17-864cbfc77066)
Slowly getting there.

GeoDerp commented:
bulk of configuration page done. Just need to test and add redundancy.
Later Ill make a video and/or some diagrams the explain the approaches of running EMHASS in docker/Home Assistant.
![Screenshot 2024-08-27 at 22-21-15 EMHASS Energy Management Optimization for Home Assistant](https://github.com/user-attachments/assets/586aec2e-8616-404e-8014-2dd64e0b5e65)

GeoDerp commented:
Greetings @davidusb-geek, hope you are doing well.

Here is a video for an overview of the PR.

If you approve, I would appreciate it if you could reply to this post with a thumbs up. Then ill send the vid to LinkedIn & Community form.

The video isn't amazing, but I hope it demonstrates the PR well enough. *I may polish the video a little more before I send it out giving time.*

Regards.

[![EMHASS - PR #334 - Standalone and Add-on merge](http://img.youtube.com/vi/T85DAdXnGFY/0.jpg)](http://www.youtube.com/watch?v=T85DAdXnGFY " EMHASS - PR #334 - Standalone and Add-on merge")

davidusb-geek commented:
Hi @GeoDerp,
I've just now had the time to check on all this. I just saw the video.
Great work as always this will be really helful as you said in the introductory part of the video.
To be honest I totally understood the motivation but did not had the time until now to really understand the work that you have done. That graph summarize it very well and it is just great to have just one unified interface to input the configuration. So big thumbs up for all this.

davidusb-geek commented:
No problem for me about migrating to github registry.

davidusb-geek commented:
One thing on my side will be to completely understand the workflow of something that happens all the time, which is the need to add a new controlling parameter in the configuration. So with all these changes I will to get used to this new workflow and understad what is needed to add (or remove) a new parameter

GeoDerp commented:
> One thing on my side will be to completely understand the workflow of something that happens all the time, which is the need to add a new controlling parameter in the configuration. So with all these changes I will to get used to this new workflow and understad what is needed to add (or remove) a new parameter

Ah yeah good point, So the steps of appending a new parameter would be: 

*Example parameter = `this_parameter_is_amazing`*

Append a line into `associations.csv` :
*So that build_params() knows what config catagorie to allocate the parameter*
```csv
...
retrieve_hass_conf,,his_parameter_is_amazing
```
 - Alternatively if you want to support this parameter with the yaml conversion *(Ie. allow the parameter to be converted from config_emhass.yaml)*
    ```csv
    ...
    retrieve_hass_conf,his_parameter_is_amazing,his_parameter_is_amazing
    ```

Append a line into the `config_defaults.json`
*To set a default value for the user if none is provided in `config.json`*
```json
"...": "...",
"this_parameter_is_amazing": [0.1,0.1]
```

Lastly, to support the configuration website to generate the parameter in the list view, append the `param_definitions.json` file:
```json
"this_parameter_is_amazing": {
      "friendly_name": "This parameter is amazing",
      "Description": "This parameter functions as you expect. It makes EMHASS AMAZING!",
      "input": "array.float",
      "default_value": 0.777
    }
```
*Note: The `default_value` in this case acts (or should act) as last resort fallback if default_config.json is not found. It also acts as the default value when you append (press plus) to an array.\* parameter*

davidusb-geek commented:
Thanks for that detailed answer. It seems clear for now but of course I will need to test that workflow.
I'll let you tell me when this PR is ready to merge.
And thank again for this! This will solve a lot of the difficulties when configuring EMHASS for the first time ;-)

davidusb-geek commented:
Hi, from the TODO list above I understand that there are still some tests to be implemented + documentation, is that right?

davidusb-geek commented:
Also CodeQL is complaining, I'm not sure if we should dismiss those

GeoDerp commented:
> Hi, from the TODO list above I understand that there are still some tests to be implemented + documentation, is that right?

That's correct. The unittests aren't necessary but the documentation kinda need a rewrite. Focusing on the new methods of running EMHASS and the parameter naming conventions.

GeoDerp commented:
> Also CodeQL is complaining, I'm not sure if we should dismiss those

I'll have a re-review of them, from my previous observations the complaints aren't anything new, from how EMHASS used to operate. I just change the methods of calling the secrets.

GeoDerp commented:
Ill see how mutch documentation I can complete tonight. Probably mainly focus on the parameter conversion. With the documentation to run EMHASS to do later be me or someone else.

davidusb-geek commented:
Hi @GeoDerp, shouldd we proceed to merge or not yet?

GeoDerp commented:
> Hi @GeoDerp, should we proceed to merge or not yet?

@davidusb-geek I'll see if I can smash out the bulk of the documentation today. Then we can merge and fine-tune it later. 👍

GeoDerp commented:
Alright, I believe the bulk of the documentation is complete. Would likely need to refine it later. Didnt manage to get to the `Emhass-Add-on` docs yet.

davidusb-geek commented:
Ok I will wait for you to complete the add-on docs part and then merge

GeoDerp commented:
> Ok I will wait for you to complete the add-on docs part and then merge

@GeoDerp , this should be done by the end of today 👍

GeoDerp commented:
Bulk of the Add-on is done 👍

GeoDerp commented:
Looks like one of those tests that fails randomly. Maybe related to daylight saving time 🤔

GeoDerp commented:
@davidusb-geek, could you have a look at f754cf9 , this is just my simple "fix" for the DST issue.
However I wouldn't be surprised if this is implemented wrong and snuffles the time incorrectly. *(Feel free to revert)* Besides that feel free to test both repos and merge.

davidusb-geek commented:
> @davidusb-geek, could you have a look at [f754cf9](https://github.com/davidusb-geek/emhass/commit/f754cf99758f2b9869b2cbfa37a929e2eff2b933) , this is just my simple "fix" for the DST issue. However I wouldn't be surprised if this is implemented wrong and snuffles the time incorrectly. _(Feel free to revert)_ Besides that feel free to test both repos and merge.

Maybe we can check but at the very beginning of the code all timestamps were handled like you have done, using UTC, which seems to me as the best option to deal with DST changes. But then this was shifted to other methods after failures on DST events. That's an open issue so let's try this and see how it goes, I have reviewed your changes an it seems all logic to me.

I will merge and test. If everything goes fine I will release a new version hopefully solving DST failures. The DST change is due for tomorrow night!

GeoDerp commented:
Amazing! Let me know if you need any clarification on anything, or need me to fix something. 👍

davidusb-geek commented:
> Amazing! Let me know if you need any clarification on anything, or need me to fix something. 👍

Ok will do 👍

A quick first question: on the add-on repo all github workflows are commented out. So this is normal since all will be dealed here in the main EMHASS code repo right?
If so I'v just disabled actions on the add-on repo for now

GeoDerp commented:
> > Amazing! Let me know if you need any clarification on anything, or need me to fix something. 👍
> 
> Ok will do 👍
> 
> A quick first question: on the add-on repo all github workflows are commented out. So this is normal since all will be dealed here in the main EMHASS code repo right?
> If so I've just disabled actions on the add-on repo for now

Yep, that is correct. 👍

davidusb-geek commented:
Ok, released a new version.
Then locally tested the add-on on my own HA instance using the method: "Adding EMHASS-Add-on into pre-existing Home Assistant environment".
Configured my `config.json` from the webserver, saved and launched a day-ahead optimization.
Everuthing went smoothly.
Proceeding to push the new add-on release version.
Impressive work @GeoDerp, Cheers!

purcell-lab commented:
> Then locally tested the add-on on my own HA instance using the method: "Adding EMHASS-Add-on into pre-existing Home Assistant environment".

I seem to be missing the upgrade step.

I took my existing 0.10.6 and upgraded to 0.11.0, but wasn't offered an opportunity to bring across my old configuration and then was running 0.11.0 with a 'default' configuration.

purcell-lab commented:
I will continue to report my experiences here, but let me know if these would be better to raise as issues.

I have now manually taken my YAML config from 0.10.6 and copy/ converted to 0.11.0, which seems to convert correctly.

There is a hard coded reference to sensor_power_photovoltaics (in the replace NAN values), but I have changed by sensor entity to sensor.apf_generation_entity, which then causes an error.

![Screenshot 2024-10-26 15 47 05](https://github.com/user-attachments/assets/48a74694-a5ff-4c6a-9aad-cc3bec2021da)
![Screenshot 2024-10-26 15 46 39](https://github.com/user-attachments/assets/ba056cdb-c55b-450e-a2da-69f59b9a3a72)

When I updated the replace NAN values with my updated entity, I could go to the next stage.

davidusb-geek commented:
> I will continue to report my experiences here, but let me know if these would be better to raise as issues.
> 
> I have now manually taken my YAML config from 0.10.6 and copy/ converted to 0.11.0, which seems to convert correctly.
> 
> There is a hard coded reference to sensor_power_photovoltaics (in the replace NAN values), but I have changed by sensor entity to sensor.apf_generation_entity, which then causes an error.
> 
> When I updated the replace NAN values with my updated entity, I could go to the next stage.

Yes it is a manual conversion of the configuration file, is not automatic.
So after manually changing to the correct names for your sensors everything is working fine?

GeoDerp commented:
> I will continue to report my experiences here, but let me know if these would be better to raise as issues.
> 
> I have now manually taken my YAML config from 0.10.6 and copy/ converted to 0.11.0, which seems to convert correctly.
> 
> There is a hard coded reference to sensor_power_photovoltaics (in the replace NAN values), but I have changed by sensor entity to sensor.apf_generation_entity, which then causes an error.
> 
> ![Screenshot 2024-10-26 15 47 05](https://github.com/user-attachments/assets/48a74694-a5ff-4c6a-9aad-cc3bec2021da)
> ![Screenshot 2024-10-26 15 46 39](https://github.com/user-attachments/assets/ba056cdb-c55b-450e-a2da-69f59b9a3a72)
> 
> When I updated the replace NAN values with my updated entity, I could go to the next stage.

Ah yes, because `sensor_linear_interp`, `sensor_replace_zero` are new parameters for the add-on users so there was no conversation. May be good to note that.

purcell-lab commented:
> So after manually changing to the correct names for your sensors everything is working fine?

No unfortunately not, I haven't been able to get a stable optimisation. When I call my MPC optimisation the call times out. As does the day ahead optimisation.

I'm going to try and simplify my. MPC call to see if I can start with something functional.

GeoDerp commented:
> > So after manually changing to the correct names for your sensors everything is working fine?
> 
> No unfortunately not, I haven't been able to get a stable optimisation. When I call my MPC optimisation the call times out. As does the day ahead optimisation.
> 
> I'm going to try and simplify my. MPC call to see if I can start with something functional.

Feel free to send the logs over if you find anything and I'll look Into it. If you could also uninstall the add-on and re-install to see if that changes anything. I'm guessing not but it might be a weird cache issue?

GeoDerp commented:
Im looking into this issue now.

GeoDerp commented:
Sorry to bug you gain @purcell-lab , while I'm fixing the other bugs, are there any logs I can observe for your issues?

kcoffau commented:
Hi,

ive migrated my config across and it runs in the UI.

Like @purcell-lab I am using the script to publish my Solcast, price and battery data from HA into EMHASS.

My first note is that my battery used to be sent as 51% as "0.51".   Its seems to now be expecting "51".  So I updated that..  But now its publishing "5105.23" back, so there seems to be a multiplier issue.

I am getting very abnormal results elsewhere as well, but i am going to start with Battery SOC.

Kel.

![image](https://github.com/user-attachments/assets/39e993c2-1815-4b21-9795-d073cce72cc0)
This is a 2 day outlook and i would expect it to cycle the battery 4 times in 2 days for my setup.

I`ll keep digging later, got to get back to "actual work", but the UI is a great step in the right direction.  The solar/inverter would be great as a searchable selector/dropdown as people find the underscore stuff confusing.



The EMHASS Logs are here:
2024-10-29 11:43:17,622 - web_server - INFO - Passed runtime parameters: {'load_cost_forecast': [0.0, 0.0, 0.0, 0.0, 0.0, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.08, 0.08, 0.08, 0.08, 0.08, 0.08, 0.08, 0.08, 0.08, 0.08, 0.08, 0.08, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.08, 0.08, 0.08, 0.08, 0.08, 0.08, 0.08, 0.08, 0.08, 0.08, 0.08, 0.08, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.0], 'prod_price_forecast': [0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03], 'pv_power_forecast': [4500, 4652, 4685, 4611, 4188, 3776, 3450, 3179, 2806, 2341, 1770, 1132, 575, 157, 110, 20, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 51, 381, 1225, 1986, 2743, 3532, 4150, 4685, 5000, 5000, 5000, 5000, 5000, 5000, 4786, 4482, 4272, 4007, 3525, 3021, 2404, 1751, 1124, 579, 168, 115, 19, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 165, 940, 1757, 2483, 3229, 3938, 4418, 4781, 5000, 5000, 5000], 'prediction_horizon': 96, **'soc_init': 51.0,** 'def_total_hours': [0]}
2024-10-29 11:43:17,623 - web_server - INFO -  >> Setting input data dict
2024-10-29 11:43:17,634 - web_server - INFO - Setting up needed data
2024-10-29 11:43:17,639 - web_server - INFO - Retrieve hass get data method initiated...
2024-10-29 11:43:20,272 - web_server - INFO - EMHASS server online, serving index.html...
2024-10-29 11:43:24,081 - web_server - INFO - EMHASS server online, serving index.html...
2024-10-29 11:43:24,390 - web_server - INFO - Retrieving weather forecast data using method = list
2024-10-29 11:43:24,393 - web_server - INFO - Retrieving data from hass for load forecast using method = naive
2024-10-29 11:43:24,394 - web_server - INFO - Retrieve hass get data method initiated...
2024-10-29 11:43:28,090 - web_server - INFO - EMHASS server online, serving index.html...
2024-10-29 11:43:36,040 - web_server - INFO -  >> Performing naive MPC optimization...
2024-10-29 11:43:36,041 - web_server - INFO - Performing naive MPC optimization
2024-10-29 11:43:36,130 - web_server - INFO - Perform an iteration of a naive MPC controller
2024-10-29 11:43:36,724 - web_server - INFO - Status: Infeasible
2024-10-29 11:43:36,725 - web_server - INFO - Total value of the Cost function = 256.80
2024-10-29 11:43:42,301 - web_server - INFO - Passed runtime parameters: {}
2024-10-29 11:43:42,301 - web_server - INFO -  >> Setting input data dict
2024-10-29 11:43:42,301 - web_server - INFO - Setting up needed data
2024-10-29 11:43:42,311 - web_server - INFO -  >> Publishing data...
2024-10-29 11:43:42,323 - web_server - INFO - Publishing data to HASS instance
2024-10-29 11:43:42,369 - web_server - INFO - Successfully posted to sensor.p_pv_forecast = 4113.88
2024-10-29 11:43:42,401 - web_server - INFO - Successfully posted to sensor.p_load_forecast = 2147.05
2024-10-29 11:43:42,424 - web_server - INFO - Successfully posted to sensor.p_deferrable0 = 0.0
2024-10-29 11:43:42,447 - web_server - INFO - Successfully posted to sensor.p_deferrable1 = 0.0
2024-10-29 11:43:42,473 - web_server - INFO - Successfully posted to sensor.p_batt_forecast = -8000.0
2024-10-29 11:43:42,497 - web_server - INFO - Successfully posted to **sensor.soc_batt_forecast = 5105.23**
2024-10-29 11:43:42,521 - web_server - INFO - Successfully posted to sensor.p_grid_forecast = 6033.17
2024-10-29 11:43:42,541 - web_server - INFO - Successfully posted to sensor.total_cost_fun_value = -124.98
2024-10-29 11:43:42,561 - web_server - INFO - Successfully posted to sensor.optim_status = Infeasible
2024-10-29 11:43:42,591 - web_server - INFO - Successfully posted to sensor.unit_load_cost = 0.0
2024-10-29 11:43:42,614 - web_server - INFO - Successfully posted to sensor.unit_prod_price = 0.03

kcoffau commented:
![image](https://github.com/user-attachments/assets/2c66158e-c16e-4583-9342-5aef5be1fb4c)
Unsure if  v0.11.1  fixed it, or what, but i am not back to normal.


i also fiddled with settings.
![image](https://github.com/user-attachments/assets/32bc3554-584d-4914-b86c-6cabc97926a4)

![image](https://github.com/user-attachments/assets/7cba7660-9f8c-4606-bb26-26f78f86b892)

Log:
2024-10-29 13:26:54,289 - web_server - INFO - Saved parameters from webserver
2024-10-29 13:27:02,395 - web_server - INFO - Passed runtime parameters: {'load_cost_forecast': [0.0, 0.0, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.08, 0.08, 0.08, 0.08, 0.08, 0.08, 0.08, 0.08, 0.08, 0.08, 0.08, 0.08, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.08, 0.08, 0.08, 0.08, 0.08, 0.08, 0.08, 0.08, 0.08, 0.08, 0.08, 0.08, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.4202, 0.0, 0.0, 0.0, 0.0], 'prod_price_forecast': [0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03, 0.03], 'pv_power_forecast': [4968, 4749, 4447, 4031, 3558, 3013, 2405, 1759, 1133, 596, 175, 116, 20, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 54, 432, 1297, 2071, 2844, 3670, 4287, 4703, 5000, 5000, 5000, 5000, 5000, 5000, 4850, 4542, 4346, 4007, 3525, 3021, 2404, 1772, 1121, 554, 151, 104, 19, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 150, 874, 1697, 2429, 3188, 3932, 4418, 4781, 5000, 5000, 5000, 5000, 5000, 4956], 'prediction_horizon': 96, **'soc_init': 0.65,** 'def_total_hours': [0]}
2024-10-29 13:27:02,396 - web_server - INFO -  >> Setting input data dict
2024-10-29 13:27:02,396 - web_server - INFO - Setting up needed data
2024-10-29 13:27:02,402 - web_server - INFO - Retrieve hass get data method initiated...
2024-10-29 13:27:05,192 - web_server - INFO - Retrieving weather forecast data using method = list
2024-10-29 13:27:05,202 - web_server - INFO - Retrieving data from hass for load forecast using method = naive
2024-10-29 13:27:05,203 - web_server - INFO - Retrieve hass get data method initiated...
2024-10-29 13:27:17,095 - web_server - INFO -  >> Performing naive MPC optimization...
2024-10-29 13:27:17,095 - web_server - INFO - Performing naive MPC optimization
2024-10-29 13:27:17,140 - web_server - INFO - Perform an iteration of a naive MPC controller
2024-10-29 13:27:18,587 - web_server - INFO - Status: Optimal
2024-10-29 13:27:18,587 - web_server - INFO - Total value of the Cost function = -8.24
2024-10-29 13:27:24,859 - web_server - INFO - Passed runtime parameters: {}
2024-10-29 13:27:24,860 - web_server - INFO -  >> Setting input data dict
2024-10-29 13:27:24,861 - web_server - INFO - Setting up needed data
2024-10-29 13:27:24,865 - web_server - INFO -  >> Publishing data...
2024-10-29 13:27:24,865 - web_server - INFO - Publishing data to HASS instance
2024-10-29 13:27:24,898 - web_server - INFO - Successfully posted to sensor.p_pv_forecast = 4997.79
2024-10-29 13:27:24,923 - web_server - INFO - Successfully posted to sensor.p_load_forecast = 1357.25
2024-10-29 13:27:24,945 - web_server - INFO - Successfully posted to sensor.p_hybrid_inverter = -3002.21
2024-10-29 13:27:24,965 - web_server - INFO - Successfully posted to sensor.p_deferrable0 = 0.0
2024-10-29 13:27:24,987 - web_server - INFO - Successfully posted to sensor.p_deferrable1 = 0.0
2024-10-29 13:27:25,008 - web_server - INFO - Successfully posted to sensor.p_batt_forecast = -8000.0
2024-10-29 13:27:25,031 - web_server - INFO - Successfully posted t**o sensor.soc_batt_forecast = 70.28**
2024-10-29 13:27:25,062 - web_server - INFO - Successfully posted to sensor.p_grid_forecast = 4359.47
2024-10-29 13:27:25,084 - web_server - INFO - Successfully posted to sensor.total_cost_fun_value = -8.24
2024-10-29 13:27:25,104 - web_server - INFO - Successfully posted to sensor.optim_status = Optimal
2024-10-29 13:27:25,127 - web_server - INFO - Successfully posted to sensor.unit_load_cost = 0.0
2024-10-29 13:27:25,152 - web_server - INFO - Successfully posted to sensor.unit_prod_price = 0.03
2024-10-29 13:27:58,844 - web_server - INFO - EMHASS server online, serving index.html...
2024-10-29 13:28:48,179 - web_server - INFO - serving configuration.html...
2024-10-29 13:28:48,320 - web_server - INFO - Obtaining parameters from config.json:

purcell-lab commented:
Upgraded to v0.11.1.

Still not having joy.

The REST post timeouts for me, I have extended timeout to 120 seconds.

![image](https://github.com/user-attachments/assets/1a7e440f-9056-452c-b089-0860fd428032)

Unfortunately the logs don't provide much context:

```logs
2024-10-29 12:39:56,310 - web_server - INFO - Launching the emhass webserver at: http://0.0.0.0:5000
2024-10-29 12:39:56,310 - web_server - INFO - Home Assistant data fetch will be performed using url: http://supervisor/core/api
2024-10-29 12:39:56,310 - web_server - INFO - The data path is: /app/data
2024-10-29 12:39:56,311 - web_server - INFO - The logging is: INFO
2024-10-29 12:39:56,314 - web_server - INFO - Using core emhass version: 0.11.1
waitress   INFO  Serving on http://0.0.0.0:5000
2024-10-29 12:40:21,401 - web_server - INFO - Passed runtime parameters: {'load_cost_forecast': [0.04, 0.05, 0.05, 0.05, 0.06, 0.08, 0.09, 0.26, 0.29, 0.33, 0.39, 0.46, 0.48, 0.45, 0.41, 0.4, 0.39, 0.26, 0.25, 0.24, 0.23, 0.22, 0.22, 0.21, 0.21, 0.21, 0.21, 0.21, 0.21, 0.21, 0.21, 0.21, 0.21, 0.22, 0.22, 0.18, 0.13, 0.11, 0.11, 0.1, 0.1, 0.07, 0.06, 0.06, 0.06, 0.05, 0.05, 0.05, 0.05, 0.05], 'prod_price_forecast': [-0.04, -0.04, -0.04, -0.03, -0.03, -0.01, 0.0, 0.02, 0.05, 0.08, 0.14, 0.2, 0.22, 0.19, 0.16, 0.15, 0.14, 0.13, 0.12, 0.11, 0.11, 0.1, 0.09, 0.09, 0.09, 0.09, 0.09, 0.09, 0.09, 0.08, 0.08, 0.09, 0.09, 0.1, 0.1, 0.06, 0.02, -0.0, -0.0, -0.01, -0.01, -0.02, -0.03, -0.03, -0.03, -0.03, -0.03, -0.03, -0.03, -0.03], 'prediction_horizon': 48, 'alpha': 0, 'beta': 1, 'num_def_loads': 6, 'def_total_hours': [0, 0, 0, 7, 13, 4], 'def_end_timestep': [0, 0, 0, 0, 0, 0], 'P_deferrable_nom': [1000, 6333, 11520, 500, 600, 11520], 'def_current_state': [0, 0, 0, 1, 1, 1], 'def_start_penalty': [1, 1, 1, 1, 1, 1], 'treat_def_as_semi_cont': [1, 1, 0, 0, 1, 0], 'set_def_constant': [0, 0, 0, 0, 0, 0], 'soc_init': 0.92, 'soc_final': 1}
2024-10-29 12:40:21,402 - web_server - INFO -  >> Setting input data dict
2024-10-29 12:40:21,402 - web_server - INFO - Setting up needed data
2024-10-29 12:40:21,454 - web_server - INFO - Retrieve hass get data method initiated...
2024-10-29 12:40:22,706 - web_server - INFO - Retrieving weather forecast data using method = scrapper
2024-10-29 12:40:26,342 - web_server - INFO - Retrieving data from hass for load forecast using method = naive
2024-10-29 12:40:26,343 - web_server - INFO - Retrieve hass get data method initiated...
2024-10-29 12:40:29,641 - web_server - INFO -  >> Performing naive MPC optimization...
2024-10-29 12:40:29,642 - web_server - INFO - Performing naive MPC optimization
2024-10-29 12:40:29,674 - web_server - INFO - Perform an iteration of a naive MPC controller
```

GeoDerp commented:
Thanks @kcoffau & @purcell-lab , I'll have a look at this and see if I can find anything. It may get out of the realm for me to fix (depending on where the issue resides). So I may need @davidusb-geek 's help. But I'll let you know after I have a look.

GeoDerp commented:
@purcell-lab , It looks like the optimization is being stuck after the last Validated window. I'm not quite sure why. Any ideas @davidusb-geek ?

```log
waitress   INFO  Serving on http://0.0.0.0:5000
2024-10-29 11:38:58,301 - web_server - INFO - EMHASS server online, serving index.html...
2024-10-29 11:39:01,586 - web_server - INFO - Passed runtime parameters: {}
2024-10-29 11:39:01,586 - web_server - INFO -  >> Setting input data dict
2024-10-29 11:39:01,586 - web_server - INFO - Setting up needed data
2024-10-29 11:39:01,599 - web_server - INFO - Retrieving weather forecast data using method = scrapper
2024-10-29 11:39:03,484 - web_server - INFO - Retrieving data from hass for load forecast using method = naive
2024-10-29 11:39:03,485 - web_server - INFO - Retrieve hass get data method initiated...
2024-10-29 11:39:03,918 - web_server - INFO -  >> Performing dayahead optimization...
2024-10-29 11:39:03,918 - web_server - INFO - Performing day-ahead forecast optimization
2024-10-29 11:39:03,920 - web_server - INFO - Perform optimization for the day-ahead
2024-10-29 11:39:04,014 - web_server - DEBUG - Deferrable load 0: Proposed optimization window: 0 --> 0
2024-10-29 11:39:04,014 - web_server - DEBUG - Deferrable load 0: Validated optimization window: 0 --> 0
2024-10-29 11:39:04,027 - web_server - DEBUG - Deferrable load 1: Proposed optimization window: 0 --> 0
2024-10-29 11:39:04,027 - web_server - DEBUG - Deferrable load 1: Validated optimization window: 0 --> 0
2024-10-29 11:39:04,038 - web_server - DEBUG - Deferrable load 2: Proposed optimization window: 0 --> 0
2024-10-29 11:39:04,038 - web_server - DEBUG - Deferrable load 2: Validated optimization window: 0 --> 0
2024-10-29 11:39:04,049 - web_server - DEBUG - Deferrable load 3: Proposed optimization window: 0 --> 0
2024-10-29 11:39:04,049 - web_server - DEBUG - Deferrable load 3: Validated optimization window: 0 --> 0
2024-10-29 11:39:04,063 - web_server - DEBUG - Deferrable load 4: Proposed optimization window: 0 --> 0
2024-10-29 11:39:04,063 - web_server - DEBUG - Deferrable load 4: Validated optimization window: 0 --> 0
2024-10-29 11:39:04,074 - web_server - DEBUG - Deferrable load 5: Proposed optimization window: 0 --> 0
2024-10-29 11:39:04,075 - web_server - DEBUG - Deferrable load 5: Validated optimization window: 0 --> 0
```

davidusb-geek commented:
> @purcell-lab , It looks like the optimization is being stuck after the last Validated window. I'm not quite sure why. Any ideas @davidusb-geek ? It looks like a battery problem. ("set_use_battery": false, fixes the issue) May be related to @kcoffau 's issue?

Unfortunately not idea for now, this will need debugging

purcell-lab commented:
Is the documentation for 0.10.6 still available, as a number of the variables are changed I still need access to the previous documentation.

On a site with 0.10.6 installed the documentation link goes to the latest version, which of course is now 0.11.1, which isn't relevant.

purcell-lab commented:
> P.s. Ill be revisiting the runtime parameters soon from this PR. I believe there are defiantly places I can fix up.

It would also be great to get more debugging info as the various states are processed, so we can see where the system need attention.

#350

GeoDerp commented:
I'll have a relook at your issue @purcell-lab likely today with 0.11.2 to see if the issue still persists.

GeoDerp commented:
> Is the documentation for 0.10.6 still available, as a number of the variables are changed I still need access to the previous documentation.
> 
> On a site with 0.10.6 installed the documentation link goes to the latest version, which of course is now 0.11.1, which isn't relevant.

In read the docs, there is a option to select version. So there may be a way for us to set like a 0.10.6 snapshot.

For now you can view the GitHub snapshots: 

https://github.com/davidusb-geek/emhass/tree/v0.10.6

https://github.com/davidusb-geek/emhass-add-on/tree/v0.10.6

GeoDerp commented:
Looks like the issue still persists with @purcell-lab 's config. I believe its something to do with the battery, I'll do some searching. Unable to find the fault so far.

kcoffau commented:
@purcell-lab 

ive compared our config:
Mine is below for reference.

I note the following differences (that just arent general settings)
**"battery_maximum_state_of_charge": 1,**
 - I had an issue with setting 1, and had to set to 0.99
**"battery_minimum_state_of_charge": 0,**
- I had an issue with setting 0, and had to set 0.01
**"set_battery_dynamic": true,"**
- i could never get this to work. Did you want to try disabling for troubleshooting?
**"weight_battery_charge": 1,"
"weight_battery_discharge": 0.05"**
- both of these values after the last release i was getting very obscure behaviour.  I set them both to "0" and it resolved it.


Ohh, and your SOC_final in your CURL is also 1.  Did you want to troubleshoot with a 0.99?


Otherwise your config looks within reason the same as mine, +/- inverter/grid/solar settings.








`{
  "battery_charge_efficiency": 0.99,
  "battery_charge_power_max": 8000,
  "battery_discharge_efficiency": 0.99,
  "battery_discharge_power_max": 8000,
  "battery_dynamic_max": 0.9,
  "battery_dynamic_min": -0.9,
  "battery_maximum_state_of_charge": 0.99,
  "battery_minimum_state_of_charge": 0.01,
  "battery_nominal_energy_capacity": 75000,
  "battery_target_state_of_charge": 0.98,
  "compute_curtailment": false,
  "continual_publish": false,
  "costfun": "profit",
  "delta_forecast_daily": 2,
  "end_timesteps_of_each_deferrable_load": [
    0,
    0
  ],
  "historic_days_to_retrieve": 3,
  "inverter_is_hybrid": true,
  "load_cost_forecast_method": "hp_hc_periods",
  "load_forecast_method": "naive",
  "load_negative": false,
  "load_offpeak_hours_cost": 0.08,
  "load_peak_hour_periods": {
    "period_hp_1": [
      {
        "start": "06:00"
      },
      {
        "end": "11:00"
      }
    ],
    "period_hp_2": [
      {
        "start": "14:00"
      },
      {
        "end": "23:59"
      }
    ]
  },
  "load_peak_hours_cost": 0.48,
  "logging_level": "INFO",
  "lp_solver": "COIN_CMD",
  "lp_solver_path": "/usr/bin/cbc",
  "maximum_power_from_grid": 15000,
  "maximum_power_to_grid": 5000,
  "method_ts_round": "nearest",
  "modules_per_string": [
    11
  ],
  "nominal_power_of_deferrable_loads": [
    0,
    0
  ],
  "number_of_deferrable_loads": 2,
  "operating_hours_of_each_deferrable_load": [
    0,
    0
  ],
  "optimization_time_step": 30,
  "photovoltaic_production_sell_price": 0.03,
  "production_price_forecast_method": "constant",
  "pv_inverter_model": [
    "Fronius_International_GmbH__Fronius_Primo_5_0_1_208_240__240V_"
  ],
  "pv_module_model": [
    "CSUN_Eurasia_Energy_Systems_Industry_and_Trade_CSUN295_60M"
  ],
  "sensor_linear_interp": [
    "sensor.power_ac_fronius_inverter_1_192_168_10_122"
  ],
  "sensor_power_load_no_var_loads": "sensor.victron_ac_l1",
  "sensor_power_photovoltaics": "sensor.power_ac_fronius_inverter_1_192_168_10_122",
  "sensor_replace_zero": [
    "sensor.power_ac_fronius_inverter_1_192_168_10_122",
    "sensor.victron_ac_l1"
  ],
  "set_battery_dynamic": false,
  "set_deferrable_load_single_constant": [
    false,
    false
  ],
  "set_deferrable_startup_penalty": [
    0,
    0
  ],
  "set_nocharge_from_grid": false,
  "set_nodischarge_to_grid": false,
  "set_total_pv_sell": false,
  "set_use_battery": true,
  "set_zero_min": false,
  "start_timesteps_of_each_deferrable_load": [
    0,
    0
  ],
  "strings_per_inverter": [
    2
  ],
  "surface_azimuth": [
    10
  ],
  "surface_tilt": [
    15
  ],
  "treat_deferrable_load_as_semi_cont": [
    true,
    true
  ],
  "weather_forecast_method": "scrapper",
  "weight_battery_charge": 0,
  "weight_battery_discharge": 0
}`

GeoDerp commented:
> @purcell-lab
> 
> ive compared our config: Mine is below for reference.
> 
> I note the following differences (that just arent general settings) **"battery_maximum_state_of_charge": 1,**
> 
>     * I had an issue with setting 1, and had to set to 0.99
> 
> 
> **"battery_minimum_state_of_charge": 0,**
> 
>     * I had an issue with setting 0, and had to set 0.01
> 
> 
> **"set_battery_dynamic": true,"**
> 
>     * i could never get this to work. Did you want to try disabling for troubleshooting?
> 
> 
> **"weight_battery_charge": 1," "weight_battery_discharge": 0.05"**
> 
>     * both of these values after the last release i was getting very obscure behaviour.  I set them both to "0" and it resolved it.
> 
> 
> Ohh, and your SOC_final in your CURL is also 1. Did you want to troubleshoot with a 0.99?
> 
> Otherwise your config looks within reason the same as mine, +/- inverter/grid/solar settings.
> 
> ```
> 
> {
>   "battery_charge_efficiency": 0.99,
>   "battery_charge_power_max": 8000,
>   "battery_discharge_efficiency": 0.99,
>   "battery_discharge_power_max": 8000,
>   "battery_dynamic_max": 0.9,
>   "battery_dynamic_min": -0.9,
>   "battery_maximum_state_of_charge": 0.99,
>   "battery_minimum_state_of_charge": 0.01,
>   "battery_nominal_energy_capacity": 75000,
>   "battery_target_state_of_charge": 0.98,
>   "compute_curtailment": false,
>   "continual_publish": false,
>   "costfun": "profit",
>   "delta_forecast_daily": 2,
>   "end_timesteps_of_each_deferrable_load": [
>     0,
>     0
>   ],
>   "historic_days_to_retrieve": 3,
>   "inverter_is_hybrid": true,
>   "load_cost_forecast_method": "hp_hc_periods",
>   "load_forecast_method": "naive",
>   "load_negative": false,
>   "load_offpeak_hours_cost": 0.08,
>   "load_peak_hour_periods": {
>     "period_hp_1": [
>       {
>         "start": "06:00"
>       },
>       {
>         "end": "11:00"
>       }
>     ],
>     "period_hp_2": [
>       {
>         "start": "14:00"
>       },
>       {
>         "end": "23:59"
>       }
>     ]
>   },
>   "load_peak_hours_cost": 0.48,
>   "logging_level": "INFO",
>   "lp_solver": "COIN_CMD",
>   "lp_solver_path": "/usr/bin/cbc",
>   "maximum_power_from_grid": 15000,
>   "maximum_power_to_grid": 5000,
>   "method_ts_round": "nearest",
>   "modules_per_string": [
>     11
>   ],
>   "nominal_power_of_deferrable_loads": [
>     0,
>     0
>   ],
>   "number_of_deferrable_loads": 2,
>   "operating_hours_of_each_deferrable_load": [
>     0,
>     0
>   ],
>   "optimization_time_step": 30,
>   "photovoltaic_production_sell_price": 0.03,
>   "production_price_forecast_method": "constant",
>   "pv_inverter_model": [
>     "Fronius_International_GmbH__Fronius_Primo_5_0_1_208_240__240V_"
>   ],
>   "pv_module_model": [
>     "CSUN_Eurasia_Energy_Systems_Industry_and_Trade_CSUN295_60M"
>   ],
>   "sensor_linear_interp": [
>     "sensor.power_ac_fronius_inverter_1_192_168_10_122"
>   ],
>   "sensor_power_load_no_var_loads": "sensor.victron_ac_l1",
>   "sensor_power_photovoltaics": "sensor.power_ac_fronius_inverter_1_192_168_10_122",
>   "sensor_replace_zero": [
>     "sensor.power_ac_fronius_inverter_1_192_168_10_122",
>     "sensor.victron_ac_l1"
>   ],
>   "set_battery_dynamic": false,
>   "set_deferrable_load_single_constant": [
>     false,
>     false
>   ],
>   "set_deferrable_startup_penalty": [
>     0,
>     0
>   ],
>   "set_nocharge_from_grid": false,
>   "set_nodischarge_to_grid": false,
>   "set_total_pv_sell": false,
>   "set_use_battery": true,
>   "set_zero_min": false,
>   "start_timesteps_of_each_deferrable_load": [
>     0,
>     0
>   ],
>   "strings_per_inverter": [
>     2
>   ],
>   "surface_azimuth": [
>     10
>   ],
>   "surface_tilt": [
>     15
>   ],
>   "treat_deferrable_load_as_semi_cont": [
>     true,
>     true
>   ],
>   "weather_forecast_method": "scrapper",
>   "weight_battery_charge": 0,
>   "weight_battery_discharge": 0
> }
> ```

Thanks @kcoffau this is great. clearly there are some bugs when it comes to the battery optimization.
